### PR TITLE
Add DB Read/Write Tracking to Benchmarking Pipeline

### PR DESCRIPTION
**Pull request by shawntabrizi**
This PR is the first step toward adding full automation to the benchmarking pipeline.

Here, we modify the Bench DB to add an additional HashMap to track keys that have been read from and written to.

We are then able to run out bechmarks, and inspect this read/write tracker to see the number of times we do a read, repeat read, write, or repeat write.

This information is then stored as part of our BenchmarkResults output and displayed as part of the CLI.

Next steps for this is:

* [x] Add a whitelist of keys to ignore for counting
* [x] Add regression analysis to the reads/writes
* [ ] Output to Rust module

### CODE REVIEW

**Comment by shawntabrizi** [comment]
Here is an example of the output:

```
➜  cli git:(shawntabrizi-bench-db-tracking) ✗ ../../../target/debug/substrate benchmark --pallet balances --extrinsic transfer --raw
2020-06-18 12:48:11 💸 new validator set of size 1 has been elected via ElectionCompute::OnChain for era 0
Pallet: "balances", Extrinsic: "transfer", Lowest values: [], Highest values: [], Steps: [], Repeat: 1
u,e,extrinsic_time,storage_root_time,reads,repeat_reads,writes,repeat_writes
1,1000,724000,53000,7,10,5,0
100,1000,715000,49000,7,10,5,0
199,1000,745000,33000,7,10,5,0
298,1000,743000,35000,7,10,5,0
397,1000,743000,33000,7,10,5,0
496,1000,790000,31000,7,10,5,0
595,1000,650000,29000,7,10,5,0
694,1000,595000,27000,7,10,5,0
793,1000,593000,28000,7,10,5,0
892,1000,608000,28000,7,10,5,0
991,1000,610000,28000,7,10,5,0
1000,2,600000,101000,7,10,5,0
1000,101,602000,28000,7,10,5,0
1000,200,603000,28000,7,10,5,0
1000,299,610000,28000,7,10,5,0
1000,398,613000,27000,7,10,5,0
1000,497,604000,29000,7,10,5,0
1000,596,604000,28000,7,10,5,0
1000,695,611000,28000,7,10,5,0
1000,794,612000,29000,7,10,5,0
1000,893,648000,27000,7,10,5,0
1000,992,604000,28000,7,10,5,0

File: client/db/src/bench.rs [hunk 1]
```diff
@@ -109,6 +147,70 @@ impl<B: BlockT> BenchmarkingState<B> {
...
+		*self.read_write_tracker.borrow_mut() = Default::default();
+	}
+
+	fn add_read_key(&self, key: &[u8]) {
+		log::trace!(target: "benchmark", "Read: {}", HexDisplay::from(&key));
+
+		let mut key_tracker = self.key_tracker.borrow_mut();
+		let mut read_write_tracker = self.read_write_tracker.borrow_mut();
+
+		let maybe_tracker = key_tracker.get(&key.to_vec());
```
Comment (kianenigma) [inline client/db/src/bench.rs:162]:
```suggestion
		let maybe_tracker = key_tracker.get(key);
```

**Comment by shawntabrizi** [comment]
@kianenigma This fixes the static read/write counting we were doing for extrinsics. It should give us the data so that we can create a weight formula with a "maximum" number of reads/writes the extrinsic will do.

But it does not fix the on_initialize/on_finalize counting that we are doing and still need to do.

**Comment by shawntabrizi** [comment]
Could I also get a review on: https://github.com/paritytech/substrate/pull/6405/files

This is a PR on top of this PR to add whitelisting to to the DB Read/Write tracking

File: client/db/src/bench.rs [hunk 2]
```diff
@@ -121,10 +223,12 @@ impl<B: BlockT> StateBackend<HashFor<B>> for BenchmarkingState<B> {
 	type TrieBackendStorage = <DbState<B> as StateBackend<HashFor<B>>>::TrieBackendStorage;
 
 	fn storage(&self, key: &[u8]) -> Result<Option<Vec<u8>>, Self::Error> {
+		self.add_read_key(key);
```
Comment (cheme) [inline client/db/src/bench.rs:226]: Did you see some significant change in bench performance from this? (it adds some map query/update so I am curious about the impact).

Comment (shawntabrizi) [reply client/db/src/bench.rs:226]: @cheme There is a negligible difference in timing that I can see across two modules:

```
diff --git a/summary.json b/summary.json
index 5d8f507..6742744 100644
--- a/summary.json
+++ b/summary.json
@@ -1,26 +1,26 @@
 {
 	"balances_set_balance": {
-		"base": "27.56",
+		"base": "27.19",
 		"u": "0",
 		"e": "0"
 	},
 	"balances_set_balance_killing": {
-		"base": "35.11",
+		"base": "33.59",
 		"u": "0",
 		"e": "0"
 	},
 	"balances_transfer": {
-		"base": "73.63",
+		"base": "65.23",
 		"u": "0",
 		"e": "0"
 	},
 	"balances_transfer_best_case": {
-		"base": "39.06",
+		"base": "34.88",
 		"u": "0",
 		"e": "0"
 	},
 	"balances_transfer_keep_alive": {
-		"base": "51.4",
+		"base": "46.17",
 		"u": "0",
 		"e": "0"
 	},
@@ -61,50 +61,54 @@
 		"c": "0"
 	},
 	"collective_close_approved": {
-		"base": "65.95",
+		"base": "68.62",
 		"m": "0.226",
-		"p": "0.487",
-		"b": "0.005"
+		"p": "0.447",
+		"b": "0.004"
 	},
 	"collective_close_disapproved": {
-		"base": "51.08",
-		"m": "0.224",
-		"p": "0.475",
-		"b": "0.003"
+		"base": "45.99",
+		"m": "0.259",
+		"p": "0.447",
+		"b": "0"
 	},
 	"collective_close_early_approved": {
-		"base": "50.82",
-		"m": "0.211",
-		"p": "0.478",
-		"b": "0.008"
+		"base": "63.57",
+		"m": "0.243",
+		"p": "0.423",
+		"b": "0.002"
 	},
 	"collective_close_early_disapproved": {
-		"base": "37.21",
-		"m": "0.239",
-		"p": "0.466",
-		"b": "0.002"
+		"base": "38.98",
+		"m": "0.24",
+		"p": "0.461",
+		"b": "0.003"
 	},
 	"collective_execute": {
-		"base": "22.62",
-		"m": "0.115",
-		"b": "0.003"
+		"base": "23.09",
+		"m": "0.112",
+		"b": "0.004"
 	},
 	"collective_propose_execute": {
-		"base": "28.12",
+		"base": "28.26",
 		"m": "0.218",
-		"b": "0.003"
+		"b": "0.004"
 	},
 	"collective_propose_proposed": {
-		"base": "49.75",
-		"m": "0.105",
-		"p": "0.502",
-		"b": "0.006"
+		"base": "53.66",
+		"m": "0.111",
+		"p": "0.427",
+		"b": "0.002"
 	},
 	"collective_set_members": {
 		"base": "0",
-		"m": "20.47",
-		"n": "0.109",
-		"p": "26.29"
+		"m": "20.51",
+		"n": "0.136",
+		"p": "25.72"
+	},
+	"collective_vote": {
+		"base": "37.39",
+		"m": "0.302"
 	},
 	"collective_vote_insert": {
 		"base": "24.03",
```

Also our main DB benchmarks are done in a separate test using the real DB backend.

Comment (cheme) [reply client/db/src/bench.rs:226]: So these are on in memorydb, thanks for posting me these it interesting numbers.

File: client/db/src/bench.rs [hunk 3]
```diff
@@ -109,6 +151,86 @@ impl<B: BlockT> BenchmarkingState<B> {
 		));
 		Ok(())
 	}
+
+	fn add_whitelist_to_tracker(&self) {
```
Comment (kianenigma) [inline client/db/src/bench.rs:155]: correct me if I am wrong: by whitelisting, basically we already assume that they have been read once? there's no inherent meaning to being whitelisted other than that.

File: client/db/src/bench.rs [hunk 4]
```diff
@@ -109,6 +151,86 @@ impl<B: BlockT> BenchmarkingState<B> {
...
+		let mut key_tracker = self.key_tracker.borrow_mut();
+
+		let whitelisted = KeyTracker {
+			has_been_read: true,
+			has_been_written: true,
+		};
+
+		let whitelist = self.whitelist.borrow_mut();
```
Comment (kianenigma) [inline client/db/src/bench.rs:163]:
```suggestion
		let whitelist = self.whitelist.borrow();
```

File: client/db/src/bench.rs (hunk 3 above)
Comment (shawntabrizi) [reply client/db/src/bench.rs:155]: Yes, this key is free to read and write to, and does not count in DB tracking.

Things like BlockNumber, the Sender Account, Events, etc...

Comment (kianenigma) [reply client/db/src/bench.rs:155]: Okay, one question then: won't this cause confusion with the read-write count? Say I submit a tx that has no reads and writes. Won't the read/write count of my tx then be equal to all the whitelisted ones?

Comment (kianenigma) [reply client/db/src/bench.rs:155]: (Either way I think it is all okay, I am just trying to make it clear for myself.)

File: frame/benchmarking/src/lib.rs [hunk 5]
```diff
@@ -775,21 +779,36 @@ macro_rules! impl_benchmark {
...
 							frame_support::debug::trace!(target: "benchmark", "Start Benchmark: {:?} {:?}", name, component_value);
 							let start_extrinsic = $crate::benchmarking::current_time();
 							closure_to_benchmark()?;
 							let finish_extrinsic = $crate::benchmarking::current_time();
 							let elapsed_extrinsic = finish_extrinsic - start_extrinsic;
+							// Commit the changes to get proper write count
+							$crate::benchmarking::commit_db();
 							frame_support::debug::trace!(target: "benchmark", "End Benchmark: {} ns", elapsed_extrinsic);
+							let read_write_count = $crate::benchmarking::read_write_count();
+							frame_support::debug::trace!(target: "benchmark", "Read/Write Count {:?}", read_write_count);
```
Comment (kianenigma) [inline frame/benchmarking/src/lib.rs:795]: this logs also in wasm, which is probably not desirable I think? 

```suggestion

							frame_support::debug::native::trace!(target: "benchmark", "Read/Write Count {:?}", read_write_count);
```

**Comment by kianenigma** [comment]
>  Output to Rust module

Will this be done in this PR or a follow up? If it could become complicated, I think it is better to separate it.

**Comment by shawntabrizi** [comment]
@kianenigma i agree, I would like to keep it separate if possible, and start merging in this working first step

File: frame/benchmarking/src/lib.rs (hunk 5 above)
Comment (shawntabrizi) [reply frame/benchmarking/src/lib.rs:795]: I also want these logs to emit during wasm execution, which is the standard way to run these benchmarks. Per conversation in "dumb questions", this should have no overhead when the log flag is not included

File: client/db/src/bench.rs (hunk 3 above)
Comment (shawntabrizi) [reply client/db/src/bench.rs:155]: no, I dont increment the read/write count when I do whitelisting. So assuming nothing is actually read/written from, you will get 0 reads and 0 writes.

**Comment by shawntabrizi** [comment]
For anyone reviewing, line width error here I think is unavoidable (caused by string literal in `hex!` macro.

**Pull request merged**